[AIRFLOW-6542] add spark-on-k8s operator/hook/sensor

roitvt opened the pull request:
Add a operator and sensor for spark-on-k8s kubernetes operator by GCP https://github.com/GoogleCloudPlatform/spark-on-k8s-operator
to send sparkApplication object to kubernetes cluster then check it's state with a sensor

---
Issue link: [AIRFLOW-6542](https://issues.apache.org/jira/browse/AIRFLOW-6542)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

mik-laj commented on airflow/contrib/example_dags/example_spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,128 @@
+    volumeMounts:
+      - name: "test-volume"
+        mountPath: "/tmp"
+"""
```
Can you store content inthe file and `template_ext`, please? DAG should be simple. 
https://airflow.readthedocs.io/en/latest/_api/airflow/models/baseoperator/index.html#airflow.models.baseoperator.BaseOperator.template_ext

roitvt replied:
I saw the other Kubernetes operators put dict with Kubernetes object to be applied by the operator.
if I put the contents in a file where to put it?

mik-laj replied:
Mixing code from different technologies is very problematic.I remember there is a word that describes this problem, but I don't remember the exact name. A similar problem occurs with SQL and other languages. SQL can be used in other code, but when it is very complex people started using ORM to connect the two worlds - SQL and Objects. 

You can create a file in the directory next to the file. For example. If your Python file is in the ``airflow/contrib/example_dags`` directory, you can keep additional files in the ``airflow/contrib/example_dags`` directory

roitvt replied:
fixed it: user will provide the path for the sparkApplication yaml I put a comment with a link to spark-pi example on spark-on-k8s git

mik-laj commented on airflow/contrib/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,77 @@
+        self.namespace = namespace
+        self.kube_config = kube_config
+        self.in_cluster = in_cluster
+        if kwargs.get('xcom_push') is not None:
```
```suggestion
```
It seems to me that it is not needed. Am I right?

mik-laj commented on airflow/contrib/sensors/spark_kubernetes_sensor.py:
```diff
@@ -0,0 +1,82 @@
+        self.namespace = namespace
+        self.kube_config = kube_config
+        self.in_cluster = in_cluster
+        self.api_client = Kuberneteshook(
```
Hook should not be initialized in the constructor but in the execute method. The constructor should be as light as possible, because this causes unnecessary load. DAG is loaded many times in various components, and the hook is needed only when executing the execute method.

roitvt replied:
I initialized the hook in the constructor so the poke method that can run many times in the sensor won't need to get a new connection each time it poke's. it's ok?

mik-laj replied:
Maybe it's worth moving it to the pre_execute method?  The constructor is called by every worker repeatedly when each task is started. It really should be low cost.

roitvt replied:
there is no pre_execute method for a sensor, I moved it to the poke method

mik-laj commented on airflow/contrib/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,77 @@
+            in_cluster=self.in_cluster
+        )
+        api_client = hook.get_conn()
+        api = client.CustomObjectsApi(api_client)
```
It is not recommended to call methods from external libraries in the operator. This is the role of hook.

roitvt replied:
the hook is creating the connection, the operator asks for the customObjectsApi if I need to do it in the hook I need to write code for each Kubernetes API object. I tried to make the Kubernetes hook general for various uses not just custom object api

mik-laj replied:
Hook should allow reuse of code. This hook ensures the best integration of Airflow with an external library.  It is perfectly fine if you need to create more methods, but you will not need to create all the methods that are available in the client, because not all methods make sense in the case of Airflow e.g. no one will perform cluster administrative operations using Airflow. If you create a method, you will be able to provide additional unique improvements for Airflow, e.g. wait for the correct completion of the operation, or provide a default value for some parameters (e.g. project ID or namespace). A possible improvement that I can now imagine is storing the namespace value in the connection configuration.  A similar improvement exists in the integration of Google Cloud Platform - https://github.com/apache/airflow/blob/43dcc13/airflow/gcp/hooks/base.py#L332-L358

roitvt replied:
now that I added Kubernetes connection type the namespace can be inside the connection but that means that I'll need connection for each namespace. my point of view is connection per Kubernetes cluster and namespace definition in the operator. even if the connection has namespace included I think that the operator needs to have the ability to override.

mik-laj replied:
I think that it is worth giving the option to configure namespace both via the operator and via the connection. Look here: https://github.com/apache/airflow/blob/43dcc13/airflow/gcp/hooks/base.py#L332-L358

Namespace is not something that is created by the individual Airflow user. Namespace should be defined at the organization-level. The organization determines which servers it wants to connect to. The operator should specify only operational parameters. Sometimes using a different namespace can also be an operational activity if DAG performs organization-level activities. But the most common cases are the namespace allocation by a specific cluster administrator.

roitvt replied:
in the connection you have namespace field, the operator/sensor can still set namespace but unless they do it we will use the namespace defined in the connection, if the connection doesn't have namespace then we will use "default" namespace.

mik-laj commented on docs/operators-and-hooks-ref.rst:
```diff
@@ -1049,8 +1049,11 @@ These integrations allow you to perform various operations using various softwar
    * - `Kubernetes <https://kubernetes.io/>`__
      - :doc:`How to use <howto/operator/kubernetes>`
      -
+     - :mod:`airflow.contrib.hooks.kubernetes_hook`
```
This is probably not part of Kubernetes, but uses it, so it should be in a separate section. A similar situation is with Hadoop and Hive. Hive uses Hadoop, but Hive is not Hadoop.

roitvt replied:
kubernetes_hook is a generalized hook that can be used in various operators so I think it can stay here 
for the spark_kubernetes_operator and spark_kubernetes_sensor where do you think they should reside?

mik-laj replied:
This is a difficult question, because it requires many components to make it work. After a long thought, I think it can stay here, but I will ask other people. It can be considered as part of the Kubernetes extension, so the operator and hook can be in one place.
CC: @kaxil @nuclearpinguin @potiuk

mik-laj commented on airflow/contrib/hooks/kubernetes_hook.py:
```diff
@@ -0,0 +1,50 @@
+        self.kube_config = kube_config
+        self.in_cluster = in_cluster
+
+    def get_conn(self):
```
Can you add typehint here also?

roitvt replied:
done :)

mik-laj commented on airflow/contrib/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,75 @@
+
+class SparkKubernetesOperator(BaseOperator):
+    """
+    creates sparkapplication object in kubernetes cluster
```
```suggestion
    Creates Spark Application Object in Kubernetes cluster:
```
Can you also add link to: https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/api-docs.md#sparkapplication ?

roitvt replied:
done :)

mik-laj commented on airflow/contrib/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,75 @@
+    :param sparkapplication_object: kubernetes custom_resource_definition of sparkApplication
+    :param namespace: kubernetes namespace to put sparkApplication
+    :param kube_config: kubernetes kube_config path
+    :param in_cluster: if airflow runs inside kubernetes pod take configuration from inside the cluster.
```
I think `kube_config` and `in_cluster` should be stored in connection and not be passed by the user. These values configured by the administrator should be constant.

roitvt replied:
there is no Kubernetes connection type so I created one!!

roitvt committed:
[AIRFLOW-6542] add spark-on-k8s operator/hook/sensor

add extra fields for kubernetes connection type

add extra fields for kubernetes connection type

add extra fields for kubernetes connection type

finish kubernetes connection implementation

update if logic to decide kube config

add template_ext

remove utf8 pragma

move from contrib to providers cncf

move from contrib to providers cncf doc also

add cncf provider hooks and sensors to index

update missing tests

fixes from kaxil and ash

add tests and squash bugs

add tests and squash bugs

fix ref

move kubernetes crd logic to hook

hook fixes from Kaxil

add yaml from example dag to setup.py

fix typo caused docs build to fail

fixes from Ash

more fixes from ash

logic fix from ash

remove useless print

mik-laj commented on airflow/contrib/example_dags/example_spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,79 @@
+# yaml definition of sparkApplication for example use:
+# https://raw.githubusercontent.com/GoogleCloudPlatform/spark-on-k8s-operator/master/examples/spark-pi.yaml
+spark_application_path = ""
+with open(spark_application_path) as spark_application_yaml:
```
Is there a reason why you didn't use template_ext here?
https://github.com/apache/airflow/blob/master/airflow/models/baseoperator.py#L251-L252

roitvt replied:
you are right template_ext is much more elegant, I added it.

mik-laj commented on airflow/contrib/hooks/kubernetes_hook.py:
```diff
@@ -0,0 +1,53 @@
+            self.log.info("loading kube_config from: default file")
+        else:
+            extra_obj = self.connection.extra_dejson
+            if "in_cluster" in extra_obj.keys():
```
All keys should have a crazy prefix, otherwise it will not be compatible with our Web UI. You will not be able to add the edit box in the connection edit form. 
For example:
GCP:
https://airflow.readthedocs.io/en/latest/howto/connection/gcp.html - `extra__google_cloud_platform__`
GRPC:
`extra__grpc__`
https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/grpc_hook.py#L123-L134
You can copy the get_field method if you need them.

mik-laj commented on airflow/contrib/hooks/kubernetes_hook.py:
```diff
@@ -0,0 +1,53 @@
+            else:
+                with tempfile.NamedTemporaryFile() as temp_config:
+                    self.log.info("loading kube_config from: connection extra json")
+                    temp_config.write(self.connection.extra.encode())
```
The configuration must be saved under the dictionary key. You can't pass the entire dictionary.

mik-laj commented on airflow/contrib/hooks/kubernetes_hook.py:
```diff
@@ -0,0 +1,53 @@
+class Kuberneteshook(BaseHook):
+    """
+    creates kubernetes API connection.
+    :param conn_id: the connection to Kubernetes cluster
```
```suggestion

        :param conn_id: the connection to Kubernetes cluster
```
This is needed for the documentation to be built successfully.

mik-laj commented on airflow/contrib/hooks/kubernetes_hook.py:
```diff
@@ -0,0 +1,53 @@
+
+class Kuberneteshook(BaseHook):
+    """
+    creates kubernetes API connection.
```
```suggestion
    Creates Kubernetes API connection.
```

mik-laj commented on airflow/contrib/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,72 @@
+class SparkKubernetesOperator(BaseOperator):
+    """
+    Creates sparkApplication object in kubernetes cluster:
+        https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/api-docs.md#sparkapplication
```
```suggestion
    .. seealso::
        For more detail about Spark Application Object have a look at the reference:
        https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/api-docs.md#sparkapplication
```

mik-laj commented on airflow/contrib/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,72 @@
+        self.conn_id = conn_id
+
+    def execute(self, context):
+        self.log.info("creating sparkApplication")
```
```suggestion
        self.log.info("Creating sparkApplication")
```

mik-laj commented on airflow/contrib/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,72 @@
+                plural="sparkapplications",
+                body=self.sparkapplication_object
+            )
+            self.log.info(response)
```
```suggestion
            self.log.debug("Response: %s", response)
```
This is too much information for this login level. Data should also be provided as arguments, not as messages. The first argument specifies the message. The remaining arguments specify the arguments of the message.

mik-laj replied:
This is especially important when we have log collection engines that index messages, but we don't necessarily want to index data. The number of messages is limited and it is much more effective.

mik-laj commented on airflow/contrib/sensors/spark_kubernetes_sensor.py:
```diff
@@ -0,0 +1,80 @@
+class SparkKubernetesSensor(BaseSensorOperator):
+    """
+    checks sparkApplication object in kubernetes cluster:
+    https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/api-docs.md#sparkapplication
```
```suggestion
    .. seealso::
        For more detail about Spark Application Object have a look at the reference:
        https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/api-docs.md#sparkapplication
```

mik-laj commented on airflow/contrib/sensors/spark_kubernetes_sensor.py:
```diff
@@ -0,0 +1,80 @@
+    checks sparkApplication object in kubernetes cluster:
+    https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/api-docs.md#sparkapplication
+    :param sparkapplication_name: sparkapplication resource name
+    :type str
```
```suggestion
    :type sparkapplication_name: str
```
This is needed for the documentation to be built successfully.

mik-laj commented on airflow/contrib/sensors/spark_kubernetes_sensor.py:
```diff
@@ -0,0 +1,80 @@
+            if sparkapplication_state in self.SUCCESS_STATES:
+                self.log.info("spark application ended successfully")
+                return True
+            raise AirflowException("unknown spark application state: %s" % sparkapplication_state)
```
```suggestion
            raise AirflowException("Unknown spark application state: %s" % sparkapplication_state)
```

mik-laj commented on airflow/contrib/sensors/spark_kubernetes_sensor.py:
```diff
@@ -0,0 +1,80 @@
+            if sparkapplication_state in self.FAILURE_STATES:
+                raise AirflowException("spark application failed with state: %s" % sparkapplication_state)
+            if sparkapplication_state in self.INTERMEDIATE_STATES:
+                self.log.info("spark application is still in state: %s", sparkapplication_state)
```
```suggestion
                self.log.info("Spark application is still in state: %s", sparkapplication_state)
```

kujon commented on airflow/providers/cncf/kubernetes/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,83 @@
+
+    @apply_defaults
+    def __init__(self,
+                 sparkapplication_file: str,
```
It would be really nice if things such as environment variables and configuration, in general, could be passed in dynamically from a DAG, rather than a file.

roitvt replied:
Can you be more specific?

kujon replied:
Sure. Compare this to e.g. `KubernetesPodOperator` - there I can individually specify each parameter from the k8s pod container spec when calling operator's constructor, e.g. image, env_vars, secrets, volumes, etc. The above implementation squashes all these arguments into a static yaml manifest, meaning that I can't pass in e.g. an airflow variable or airflow connection to the spark job. If you replaced `sparkapplication_file` with all of [these arguments](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/pkg/apis/sparkoperator.k8s.io/v1beta2/types.go#L181-L276) instead, such dynamic configuration of the job would be possible.

roitvt replied:
I use template_ext so you can use Jinja templating in the yaml file.
for example: [https://github.com/apache/airflow/blob/967003ad1e7d4907b4d3046cbbe326b5c473c0f4/airflow/providers/cncf/kubernetes/example_dags/example_spark_kubernetes_operator_spark_pi.yaml#L22](url)

kujon replied:
apologies, I missed that! in such case, the only thing then is the inconsistency with the `KubernetesPodOperator`, all of the functionality is there.

kaxil commented on airflow/models/connection.py:
```diff
@@ -61,6 +61,7 @@
     "hiveserver2": ("airflow.providers.apache.hive.hooks.hive.HiveServer2Hook", "hiveserver2_conn_id"),
     "jdbc": ("airflow.providers.jdbc.hooks.jdbc.JdbcHook", "jdbc_conn_id"),
     "jira": ("airflow.providers.jira.hooks.jira.JiraHook", "jira_conn_id"),
+    "kubernetes": ("airflow.providers.cncf.kubernetes.hooks.kubernetes_hook", "kubernetes_conn_id"),
```
```suggestion
    "kubernetes": ("airflow.providers.cncf.kubernetes.hooks.kubernetes.KubernetesHook", "kubernetes_conn_id"),
```

kaxil commented on airflow/providers/cncf/kubernetes/hooks/kubernetes_hook.py:
```diff
@@ -0,0 +1,77 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
```
We should rename the file to `airflow/providers/cncf/kubernetes/hooks/kubernetes.py` like the other hooks

kaxil commented on airflow/providers/cncf/kubernetes/hooks/kubernetes_hook.py:
```diff
@@ -0,0 +1,77 @@
+from airflow.hooks.base_hook import BaseHook
+
+
+class Kuberneteshook(BaseHook):
```
```suggestion
class KubernetesHook(BaseHook):
```

ashb commented on airflow/providers/cncf/kubernetes/hooks/kubernetes_hook.py:
```diff
@@ -0,0 +1,77 @@
+                self.log.debug("loading kube_config from: connection kube_config")
+                temp_config.write(self._get_field("kube_config").encode())
+                config.load_kube_config(temp_config.name)
+                temp_config.flush()
```
There's no point flushing the file _after_ it's been read by load_kube_config. Either the flush isn't needed, or it should be called before config.load_kube_config.

Is there no option to load from a dict instead of having to write it to a temp file?

roitvt replied:
I don't like this solution either, but I dug into Kubernetes python API and didn't found a better solution. so I write the Kube config to temp file let Kubernetes python API to load it and then flush.

ashb replied:
There is zero point flushing after kube has read from it - that does nothing.

ashb commented on airflow/providers/cncf/kubernetes/hooks/kubernetes_hook.py:
```diff
@@ -0,0 +1,77 @@
+        """
+        Returns the namespace that defined in the connection
+        """
+        namespace = self._get_field("namespace")
```
```suggestion
        return self._get_field("namespace", default="default")
```

would do for this method, no?

ashb commented on airflow/providers/cncf/kubernetes/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,83 @@
+class SparkKubernetesOperator(BaseOperator):
+    """
+    Creates sparkApplication object in kubernetes cluster:
+       .. seealso::
```
```suggestion

       .. seealso::
```

ashb commented on airflow/providers/cncf/kubernetes/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,83 @@
+        For more detail about Spark Application Object have a look at the reference:
+        https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/api-docs.md#sparkapplication
+
+    :param sparkapplication_file: filepath to kubernetes custom_resource_definition of sparkApplication
```
```suggestion
    :param application_file: filepath to kubernetes custom_resource_definition of sparkApplication
```

We're already in "spark" here, We don't have to mention spark again I don't think.

ashb commented on airflow/providers/cncf/kubernetes/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,83 @@
+    :type sparkapplication_file:  str
+    :param namespace: kubernetes namespace to put sparkApplication
+    :type namespace: str
+    :param conn_id: the connection to Kubernetes cluster
```
```suggestion
    :param kubernetes_conn_id: the connection to Kubernetes cluster
```

We prefix the connection id params in operators with the _type_ of the connection (which is done to make  site-wide policies easier to implement)

ashb commented on airflow/providers/cncf/kubernetes/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,83 @@
+        self.log.info("Creating sparkApplication")
+        hook = Kuberneteshook(conn_id=self.conn_id)
+        api_client = hook.get_conn()
+        api = client.CustomObjectsApi(api_client)
```
This feels like it should be in the hook, not the operator.

roitvt replied:
but then I'll need to write code in the hook for every Kubernetes API instead of just return general API connection and the operator decide which API kind he opens.
what do you think is right?

ashb replied:
Oh sorry I was misreading this code - `client` was confusing me as a module name

ashb commented on airflow/providers/cncf/kubernetes/operators/spark_kubernetes_operator.py:
```diff
@@ -0,0 +1,83 @@
+        api_client = hook.get_conn()
+        api = client.CustomObjectsApi(api_client)
+        try:
+            sparkapplication_dict = yaml.safe_load(self.sparkapplication_file)
```
Since you have specified the template_ext, if this file exists the contents of the `sparkapplication_file` attribute will be _set to the content of the file_, not the filename. 

This needs tests! 

Edit: which won't break the operator (I'm confusing yaml with json which has load and loads methods). Still. This definitely needs tests added to the PR.

ashb commented on tests/test_project_structure.py:
```diff
@@ -36,6 +36,9 @@
     'tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py',
+    'tests/providers/cncf/kubernetes/operators/test_spark_kubernetes_operator.py',
+    'tests/providers/cncf/kubernetes/hooks/test_kubernetes_hook.py',
+    'tests/providers/cncf/kubernetes/sensors/test_spark_kubernetes_sensor.py',
```
Naughty Naughty.

roitvt replied:
I know :( unfortunately it still not possible to mock Kubernetes API in python for unit testing but I'll try to write tests to what I can.

mik-laj replied:
Why is this not possible? I think that shouldn't be a problem.

roitvt replied:
you are right! I added tests

ashb commented on airflow/providers/cncf/kubernetes/hooks/kubernetes_hook.py:
```diff
@@ -0,0 +1,77 @@
+from airflow.hooks.base_hook import BaseHook
+
+
+class Kuberneteshook(BaseHook):
```
I wonder if we could share some of the connection code we already have for both the KubernetesPodOperator and the Kubernetes Executor. Or should the KubePodOperator be slowly reworked (in a separate PR) to use this hook?

kaxil replied:
I feel like KubePodOperator should use this hook.

kaxil replied:
Should be a separate PR though

roitvt replied:
so would you like me to open new PR for the Kubernetes connection? I think that KubePodOperator can benefit from using this hook as it can make it work on remote k8s clusters.

ashb replied:
Non, this PR can stay as it is, just talking about future work.

KubePodOp already works against external clusters though

mik-laj commented on airflow/providers/cncf/kubernetes/example_dags/example_spark_kubernetes_operator_spark_pi.yaml:
```diff
@@ -0,0 +1,57 @@
+#
```
Can you add this file to MANIFEST.in? 
https://github.com/apache/airflow/blob/master/MANIFEST.in

roitvt replied:
done

mik-laj commented on airflow/providers/cncf/kubernetes/operators/spark_kubernetes.py:
```diff
@@ -0,0 +1,88 @@
+        else:
+            namespace = self.namespace
+        try:
+            response = api.create_namespaced_custom_object(
```
This logic should be in the hook so that it can be used again by other custom operators. Then we can add more methods that will start operators and wait for its completion or other depending on the situation.

roitvt replied:
done

alessioale reviewed (commented):
Hello,

your pull request seems very interesting. Currently, I am using Airflow and Spark Operator on Kubernetes separately and this PR seems a very good solution to use them together.

Dealing with the code I would like to understand some points:

1.  Why are two different tasks used for creating and checking the status of Spark Application (**SparkKubernetesOperator** and **SparkKubernetesSensor**) ?
Is it possible to merge them in a single Airflow operator ? In this way you can use only a single task in the DAG to manage a Spark Application.

2. when a Custom resource is created/watched, I noticed the following part:
```
group="sparkoperator.k8s.io",
version="v1beta2",
plural="sparkapplications",
```

Why are these variables hard-coded here instead of retrieving them from the spark application definition (yaml) ? 

Having a generic airflow operator that is able to manage all kinds of custom resources and not only the spark applications would be easier to re-use.

Thanks, 
Alessio

kaxil commented on MANIFEST.in:
```diff
@@ -33,3 +33,4 @@ graft scripts/upstart
 graft airflow/config_templates
 recursive-exclude airflow/www/node_modules *
 global-exclude __pycache__  *.pyc
+include airflow/providers/cncf/kubernetes/example_dags/example_spark_kubernetes_operator_spark_pi.yaml
```
Can you include this file here too:

https://github.com/apache/airflow/blob/master/setup.py#L455

roitvt replied:
included

kaxil commented on airflow/providers/cncf/kubernetes/hooks/kubernetes.py:
```diff
@@ -0,0 +1,148 @@
+        conn_id="kubernetes_default"
+    ):
+        self.connection = self.get_connection(conn_id)
+        self.extras = self.connection.extra_dejson
```
We should not have "connection to DB" when instantiating Hook. The main reason is DAG files are parsed at regular interval and this can cause an increase in DB connections

kaxil replied:
Similar to https://github.com/apache/airflow/blob/master/airflow/providers/microsoft/azure/hooks/azure_cosmos.py

roitvt replied:
fixed

kaxil commented on airflow/providers/cncf/kubernetes/hooks/kubernetes.py:
```diff
@@ -0,0 +1,148 @@
+                temp_config.write(self._get_field("kube_config").encode())
+                temp_config.flush()
+                config.load_kube_config(temp_config.name)
+                temp_config.close()
```
We don't need to close it explicitly as we are using File Context Manager.

So you can replace the following call:

```python
file = open('some_file', 'w')
try:
    file.write('Hola!')
finally:
    file.close()
```

as 

```python
with open('some_file', 'w') as opened_file:
    opened_file.write('Hola!')
```

roitvt replied:
fixed!

kaxil commented on airflow/providers/cncf/kubernetes/hooks/kubernetes.py:
```diff
@@ -0,0 +1,148 @@
+                                          version,
+                                          plural,
+                                          body,
+                                          namespace=None
```
Can you add Type Annotations for all

roitvt replied:
added :)

kaxil reviewed (commented):
Minor comments but mostly LGTM

kaxil reviewed (approved):
LGTM. Would love 1 other committer to review and approve it too before I merge.

cc @dimberman @mik-laj @ ashb  @Fokko

ashb commented on airflow/providers/cncf/kubernetes/sensors/spark_kubernetes.py:
```diff
@@ -0,0 +1,78 @@
+class SparkKubernetesSensor(BaseSensorOperator):
+    """
+    Checks sparkApplication object in kubernetes cluster:
+       .. seealso::
```
Have you checked how this looks once rendered? I think it is slightly in-correct:
```suggestion

    .. seealso::
```

ashb commented on airflow/providers/cncf/kubernetes/sensors/spark_kubernetes.py:
```diff
@@ -0,0 +1,78 @@
+    Checks sparkApplication object in kubernetes cluster:
+       .. seealso::
+        For more detail about Spark Application Object have a look at the reference:
+        https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/api-docs.md#sparkapplication
```
Can we point at a more fixed release/url? This URL has a high risk of breaking in the future.

roitvt replied:
https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/v1beta2-1.1.0-2.4.5/docs/api-docs.md ?

ashb replied:
https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/v1beta2-1.1.0-2.4.5/docs/api-docs.md#sparkapplication probably.

ashb commented on airflow/providers/cncf/kubernetes/sensors/spark_kubernetes.py:
```diff
@@ -0,0 +1,78 @@
+    """
+
+    template_fields = ('application_name', 'namespace')
+    INTERMEDIATE_STATES = ('SUBMITTED', 'RUNNING', 'PENDING_RERUN', '')
```
What's the `''` state here for?

roitvt replied:
the spark operator can give an empty state '' that called "new state" and after a second the state is filled with SUBMITTED, RUNNING ...

ashb commented on airflow/providers/cncf/kubernetes/sensors/spark_kubernetes.py:
```diff
@@ -0,0 +1,78 @@
+    template_fields = ('application_name', 'namespace')
+    INTERMEDIATE_STATES = ('SUBMITTED', 'RUNNING', 'PENDING_RERUN', '')
+    FAILURE_STATES = ('FAILED', 'UNKNOWN')
+    SUCCESS_STATES = 'COMPLETED'
```
```suggestion
    SUCCESS_STATES = ('COMPLETED',)
```

roitvt replied:
fixed :)

ashb commented on airflow/providers/cncf/kubernetes/sensors/spark_kubernetes.py:
```diff
@@ -0,0 +1,78 @@
+            return False
+        if application_state in self.FAILURE_STATES:
+            raise AirflowException("Spark application failed with state: %s" % application_state)
+        if application_state in self.INTERMEDIATE_STATES:
```
Do we really need a list of intermediate_states -- having a list of FAILURE and SUCCESS states should be enough, and this can just be the "else" case.

roitvt replied:
you are right, fixed :)

ashb commented on airflow/providers/cncf/kubernetes/example_dags/example_spark_kubernetes_operator_spark_pi.yaml:
```diff
@@ -0,0 +1,57 @@
+apiVersion: "sparkoperator.k8s.io/v1beta2"
+kind: SparkApplication
+metadata:
+  name: spark-pi-{{ ds }}-{{ task_instance.try_number }}
```
```suggestion
  name: "spark-pi-{{ ds }}-{{ task_instance.try_number }}"
```

Reduces the risk of this becoming invalid YAML after rendering

roitvt replied:
fixed :)

ashb commented on airflow/providers/cncf/kubernetes/operators/spark_kubernetes.py:
```diff
@@ -0,0 +1,65 @@
+    def __init__(self,
+                 application_file: str,
+                 namespace: Optional[str] = None,
+                 conn_id: str = 'kubernetes_default',
```
```suggestion
                 kubernetes_conn_id: str = 'kubernetes_default',
```

(This is so that site-wide defaults can be configured using "policy" mechanism.) Likewise for the sensor.

roitvt replied:
fixed :)

ashb commented on airflow/www/static/js/connection_form.js:
```diff
@@ -55,6 +55,10 @@ $(document).ready(function () {
       }
     },
+    kubernetes: {
+      hidden_fields: ['host', 'schema', 'login', 'password', 'port', 'extra'],
```
Why are we hiding host? I would expect this to be where I would configure the kube api server when not running in a cluster.

roitvt replied:
when not running in a cluster i want to get kube config in JSON format,
Kube config is the Kubernetes connection string.
https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/

ashb commented on airflow/providers/cncf/kubernetes/hooks/kubernetes.py:
```diff
@@ -0,0 +1,151 @@
+        extras = connection.extra_dejson
+        namespace = extras.get("extra__kubernetes__namespace")
+        print(extras)
+        return "default" if namespace is None else namespace
```
```suggestion
        namespace = extras.get("extra__kubernetes__namespace", default="default")
        return namespace
```

roitvt replied:
thanks, Ash :) I fixed it
